File: frontend/lib/features/insight/data/pattern_tip.dart
class PatternTip {
  final String patternKey;
  final String tipText;
  final DateTime generatedAt;

  const PatternTip({
    required this.patternKey,
    required this.tipText,
    required this.generatedAt,
  });

  factory PatternTip.fromJson(Map<String, dynamic> json) => PatternTip(
        patternKey: json['pattern_key'] as String,
        tipText: json['tip_text'] as String,
        generatedAt: DateTime.parse(json['generated_at'] as String),
      );
}

File: backend/app/models/pattern_tip.py
"""PatternTip — cached LLM-generated tip for a (user, pattern) combo."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, ForeignKey, String, Text, UniqueConstraint, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base

if TYPE_CHECKING:
    from app.models.user import User


class PatternTip(Base):
    __tablename__ = "pattern_tips"
    __table_args__ = (
        UniqueConstraint("user_id", "pattern_key", name="pattern_tips_user_key_unique"),
    )

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )
    pattern_key: Mapped[str] = mapped_column(String(64), nullable=False)
    tip_text: Mapped[str] = mapped_column(Text, nullable=False)
    generated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    user: Mapped[User] = relationship("User")
